### PR TITLE
automated integration tests for cloudstack

### PR DESCRIPTION
**Pull request by resmo**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
tests

##### ANSIBLE VERSION

```
2.3
```

##### SUMMARY
automated integration tests for cloudstack

/cc @gundalow @mattclay

### CODE REVIEW

**Comment by gundalow** [comment]
Hi,
Thanks for this. Not managed to catch you online at the same point as mattclay (West Coast USA)

How slow is it to bring this up in Docker? If it's only a few minutes then we can add that into Shippable, rather than requiring external resources (something we are trying to remove).

The tests will need moving from `integration/roles` to `integration/targets`

**Comment by gundalow** [comment]
@resmo When you get a chance could you have a look at my comments above.

**Comment by resmo** [comment]
@gundalow yes, moving them to targets should not be a big deal. But I am currently too busy. I should get more time next week though.

**Comment by resmo** [comment]
@gundalow @mattclay  I have done some work and have a working prove of work. 

Still room left to improve and I would need some help to integrate it properly into the new best practice of integration testing :)

Look at the cloudstack.sh to see how it works; It bascically starts a docker image and a playbook containing some roles. One role is the test_cs_prepare, in which I do all the authentication preparation. The rest is self explainable. 

On my machine it takes about 6 minutes to start the setup. I currently only run one integration test `test_cs_user` (see https://github.com/ansible/ansible/pull/20552/commits/af707bdb0b911406d6a3e387666141ca49d8f173#diff-fd35e8e13de0ec5189e30a883b8f2b3cR10)

Feedback appreciated :)

**Comment by mattclay** [comment]
@resmo I'd like to see the CloudStack tests as first class test targets, just like the rest of our integration tests. I'm working on a few ideas on how to do that cleanly.

**Comment by mattclay** [comment]
I'd like to see the Dockerfile added to our repo, the same as the others we use for testing. We can wait to do that until we get closer to having everything working.

**Comment by resmo** [comment]
we are now at 9min 10s, only for the tests (631 tasks) to run.

**Comment by mattclay** [comment]
@resmo Could you open a PR to add the `cloudstack-sim` `Dockerfile` and dependencies to `test/utils/docker/cloudstack-sim/`? I'd like to be able to build it from the Ansible Docker Hub account so I can reference the image as `ansible/ansible:cloudstack-sim` from within `ansible-test`.

**Comment by mattclay** [comment]
@resmo I've deployed the initial support for testing CloudStack modules via ansible-test. I moved over some of the integration test roles to ansible-test targets. The remaining ones may require your changes from this PR before they can be migrated.

Here's the PR with the initial migration: https://github.com/ansible/ansible/pull/24317

**Comment by resmo** [comment]
Yey, it passes :)

**Comment by mattclay** [comment]
Test roles which were copied to the targets directory should be removed from the roles directory.

**Comment by resmo** [comment]
Yes of course, cleaning up

**Comment by resmo** [comment]
* Cleaned up old roles
* Kept the cloustack.yml in integration and changed the role path so the test can be easily run locally e.g. `make cloudstack TEST_FLAGS="--tags cs_user"`
* Added the docker src for cloudstack-simulator

Anything else? ;)

**Comment by mattclay** [comment]
You can run the tests locally with: `ansible-test integration cloud/cs/ -v`

If you don't want `ansible-test` to start the simulator for you, use the `test/integration/cloud-config-cs.ini.template` to tell it where to find the test endpoint.

Does that eliminate the need to maintain the legacy playbook and `make` command for you?

**Comment by resmo** [comment]
Nice! In this case I'll clean up the Makefile. I see no need to keep it.

**Comment by mattclay** [comment]
If you do encounter issues running the tests via `ansible-test` locally, please let me know, as it's intended to be fully usable (and convenient) for both local testing and running the tests on Shippable.

File: test/integration/targets/cs_instance/tasks/main.yml [hunk 1]
```diff
@@ -9,3 +9,12 @@
...
 - include: cleanup.yml
+
+#- include: setup.yml
+#  tags: test_cs_instance_adv
+#- include: present_adv.yml
+#  tags: test_cs_instance_adv
+#- include: absent.yml
+#  tags: test_cs_instance_adv
+#- include: cleanup.yml
+#  tags: test_cs_instance_adv
```
Comment (mattclay) [inline test/integration/targets/cs_instance/tasks/main.yml]: Why are these includes commented out?

File: test/integration/targets/cs_prepare/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
```
Comment (mattclay) [inline test/integration/targets/cs_prepare/defaults/main.yml]: There shouldn't be any need for `cs_prepare` with `ansible-test`. It takes care of the requirements when running with any of the options: `--tox`, `--docker` or `--requirements`, as well as configuring the environment and fetching the API keys from the simulator.

File: test/utils/docker/cloudstack-simulator/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# CloudStack Simulator as Docker Image
```
Comment (mattclay) [inline test/utils/docker/cloudstack-simulator/README.md]: For new documentation we prefer reStructuredText instead of Markdown.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/integration/targets/cs_instance/tasks/main.yml (hunk 1 above)
Comment (resmo) [reply test/integration/targets/cs_instance/tasks/main.yml]: good question, I can't remember why. Let me verify if it works.

File: test/integration/targets/cs_prepare/defaults/main.yml (hunk 2 above)
Comment (resmo) [reply test/integration/targets/cs_prepare/defaults/main.yml]: ack, was used in legacy. removing.

File: test/integration/targets/cs_instance/tasks/main.yml (hunk 1 above)
Comment (resmo) [reply test/integration/targets/cs_instance/tasks/main.yml]: I checked again. These tests are related to features I am developing regarding advanced networking. So these tests do not reflect the current state. I'll remove them for now and add them once this is merged along with the features.

**Review by mattclay — APPROVED** [review]

**Pull request merged**